usaspending_api/common/api_request_utils.py

```diff
@@ -0,0 +1,142 @@
+        date_start[date_field + '__gte'] = self.fy_start_date
+        date_end[date_field + '__lte'] = self.fy_end_date
+        if as_dict:
+            return {**date_start, **date_end}
+        else:
+            return (Q(**date_start) & Q(**date_end))
+
+
+# This class supports multiple methods of dynamically creating filter queries
+class FilterGenerator():
+    operators = {
+        'equals': '',
+        'less_than': '__lt',
+        'greater_than': '__gt',
+        'contains': '__icontains',
+        'in': '__in',
+        'less_than_or_equal': '__lte',
+        'greather_than_or_equal': '__gte',
+        'range': '__range'
+    }
+
+    # Creating the class requires a filter map - this maps one parameter filter
+    # key to another, for instance you could map "fpds_code" to "subtier_agency__fpds_code"
+    # This is useful for allowing users to filter on a fk relationship without
+    # having to specify the more complicated filter
+    # Additionally, ignored parameters specifies parameters to ignore. Always includes
+    # to ["page", "limit"]
+    def __init__(self, filter_map={}, ignored_parameters=[]):
+        self.filter_map = filter_map
+        self.ignored_parameters = ['page', 'limit'] + ignored_parameters
+
+    # Pass in request.GET and you'll get back a **kwargs object suitable for use
+    # in .filter()
+    # NOTE: GET will really only support 'AND' of filters, to use OR we'll need
+    # a more complex request object via POST
+    def create_from_get(self, parameters):
+        return_arguments = {}
+        for key in parameters:
+            if key in self.ignored_parameters:
+                continue
+            if key in self.filter_map:
+                return_arguments[self.filter_map[key]] = parameters[key]
+            else:
+                return_arguments[key] = parameters[key]
+        return return_arguments
+
+    # Creates a Q object from a POST query. Example of a post query:
+    # {
+    #     'page': 1,
+    #     'limit': 100,
+    #     'filters': [
+    #         {
+    #             'combine_method': 'OR',
+    #             'filters': [ . . . ]
+    #         },
+    #         {
+    #             'field': <FIELD_NAME>
+    #             'operation': <OPERATION>
+    #             'value': <VALUE>
+    #         },
+    #     ]
+    # }
+    # If the 'combine_method' is present in a filter, you MUST specify another
+    # 'filters' set in that object of filters to combine
+    # The combination method for filters at the root level is 'AND'
+    # Available operations are equals, less_than, greater_than, contains, in, less_than_or_equal, greather_than_or_equal, range
+    # Note that contains is always case insensitive
+    def create_from_post(self, parameters):
+        return self.create_q_from_filter_list(parameters.get('filters', []))
+
+    def create_q_from_filter_list(self, filter_list, combine_method='AND'):
+        q_object = Q()
+        for filt in filter_list:
```

Comment:
what are your thoughts on how we should sanitize the input for different filters? Perhaps for now we could raise a 400 somewhere in this block and in the next sprint try and surface some detailed info about which filter or parameter value was malformed.